# Tarea 05 Motor Trend Car Road Tests

Nicolás Martínez Gutiérrez

18/09/2025

751746

## 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

In [12]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [13]:
data = pd.read_excel("Motor Trend Car Road Tests.xlsx")

In [14]:
y = data["mpg"]
x = data.drop(columns=["mpg", "model"])

In [15]:
n = len(data)

In [16]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.4, random_state = 137)

### Escalamiento

In [17]:
from sklearn.preprocessing import StandardScaler

In [18]:
scaler = StandardScaler().fit(x_train)
X_train_scaled = scaler.transform(x_train)

In [19]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

### Train

In [20]:
y_hat = lr.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9900805114414538

In [21]:
n = len(X_train_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_train_scaled])
ols = sm.OLS(y_train, X)
results = ols.fit()
results.summary()

C:\Users\Nicolas\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=12 observations were given.
  return hypotest_fun_in(*args, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.242
Time:                        16:03:33   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1            40.7571     54.010      0.755      0.588    -645.499     727.013
x2            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x3           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x4            14.1135     13.469      1.048      0.485    -157.022     185.249
x5            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x6             1.5727      3.475      0.453      0.729     -42.583      45.729
x7            32.4926     41.262      0.787      0.575    -491.795     556.780
x8            36.1394     42.669      0.847      0.553    -506.022     578.301
x9           -38.3551     43.625     -0.879      0.541    -592.668     515.958
x10           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                         382.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

##### Interpretación signos betas

- const = 17.4333. Si todos los predictores se fijaran en cero, el modelo estima que el auto tendría un rendimiento de aproximadamente 17.43 millas por galón.

- x1 = 40.7571. Cuando aumenta este predictor, el valor de mpg crece en 40.76 unidades, indicando una relación positiva fuerte con el rendimiento de combustible.

- x2 = -2.4635. Un incremento en esta variable hace que el mpg disminuya en 2.46 unidades, lo que refleja mayor consumo.

- x3 = -10.1581. Si este predictor aumenta, el mpg baja en 10.16 unidades, mostrando un efecto negativo importante sobre la eficiencia.

- x4 = 14.1135. Un aumento en este valor se asocia con un incremento de 14.11 unidades en el mpg, mejorando el rendimiento.

- x5 = -0.8005. Cuando crece este factor, el mpg se reduce en 0.80 unidades.

- x6 = 1.5727. Este predictor se relaciona con un incremento de 1.57 unidades en el mpg, reflejando mejor eficiencia.

- x7 = 32.4926. Al aumentar esta variable, el mpg crece en 32.49 unidades, lo que implica un efecto positivo fuerte.

- x8 = 36.1394. Si este valor sube, el mpg se incrementa en 36.14 unidades.

- x9 = -38.3551. Un aumento en esta variable reduce el mpg en 38.36 unidades, indicando un efecto muy negativo sobre la eficiencia.

- x10 = 11.9707. Este factor se asocia con un aumento de 11.97 unidades en el mpg, mostrando una relación positiva.

### Regularización L2 Train

In [22]:
l2 = Ridge(alpha = 0.1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=0.1)

In [23]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9805234796986707

In [24]:
l2.intercept_

np.float64(17.43333333333333)

In [25]:
l2.coef_

array([-2.07112439, -2.90605012,  4.04654138,  3.04446081, -0.0214042 ,
        2.34685586, -0.24193943,  2.00723997, -3.10258427, -1.39667525])

In [26]:
l2 = Ridge(alpha = 1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=1)

In [27]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9386746881405902

In [28]:
l2.intercept_

np.float64(17.433333333333334)

In [29]:
l2.coef_

array([-1.80358779, -1.2857849 ,  1.01328527,  1.67274528, -0.7201921 ,
        1.26762625,  0.01863527,  1.40125429, -1.20813325, -0.75672773])

In [30]:
l2 = Ridge(alpha = 10)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=10)

In [31]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.8637598356203311

In [32]:
l2.intercept_

np.float64(17.433333333333334)

In [33]:
l2.coef_

array([-1.03418007, -0.81857091, -0.44645182,  0.78390085, -0.76522818,
        0.53725291,  0.41802987,  0.74654923,  0.06544585, -0.48870091])

### Interpretación

Con lambda = 0.1 se obtuvo un R2 de entrenamiento de aproximadamente 0.9805, lo que indica un ajuste casi perfecto a los datos, aunque con el riesgo de sobreajuste porque los coeficientes apenas fueron penalizados. Al aumentar a lambda = 1, el R2 de entrenamiento bajó a 0.9386, lo que refleja un ajuste más moderado, pero con coeficientes más pequeños y estables, logrando un mejor equilibrio entre ajuste y generalización. Finalmente, con lambda = 10, el R2 de entrenamiento disminuyó a 0.8637, evidenciando que la penalización es mucho más fuerte, reduciendo de forma considerable los coeficientes y la capacidad explicativa, pero al mismo tiempo generando un modelo más simple y con menor riesgo de sobreajuste. En conjunto, estos resultados muestran que valores pequeños de lambda priorizan el ajuste, mientras que valores más grandes buscan estabilidad y simplicidad en el modelo, a costa de sacrificar algo de poder predictivo.

### Test

In [34]:
X_test_scaled = scaler.transform(x_test)

In [35]:
lr = LinearRegression()
lr.fit(X_test_scaled, y_test)

LinearRegression()

In [36]:
y_hat_test = lr.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.9014322110655981

In [37]:
n = len(X_test_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_test_scaled])
ols = sm.OLS(y_test, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.901
Model:                            OLS   Adj. R-squared:                  0.792
Method:                 Least Squares   F-statistic:                     8.231
Date:                Thu, 18 Sep 2025   Prob (F-statistic):            0.00201
Time:                        16:03:33   Log-Likelihood:                -39.494
No. Observations:                  20   AIC:                             101.0
Df Residuals:                       9   BIC:                             111.9
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4166      1.097     15.872      0.000      14.934      19.899
x1             1.2438      2.055      0.605      0.560      -3.406       5.894
x2             2.8445      2.990      0.951      0.366      -3.919       9.608
x3            -2.3518      2.339     -1.005      0.341      -7.643       2.940
x4             0.5595      1.067      0.525      0.613      -1.853       2.972
x5            -4.4764      2.451     -1.826      0.101     -10.022       1.069
x6             1.5293      1.587      0.964      0.360      -2.061       5.119
x7             0.5587      1.110      0.503      0.627      -1.952       3.069
x8             0.7584      1.215      0.624      0.548      -1.989       3.506
x9             2.2208      1.634      1.359      0.207      -1.476       5.917
x10           -1.0744      2.233     -0.481      0.642      -6.127       3.978
==============================================================================
Omnibus:                        7.229   Durbin-Watson:                   2.199
Prob(Omnibus):                  0.027   Jarque-Bera (JB):                4.624
Skew:                           0.963   Prob(JB):                       0.0990
Kurtosis:                       4.356   Cond. No.                         25.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpretación signos betas

- Const = 17.4166. Representa el valor base del mpg, es decir, el rendimiento esperado cuando todas las variables predictoras son cero. El modelo estima un consumo aproximado de 17.42 millas por galón en ese caso.

- x1 = 1.2438. A mayor número de cilindros, el mpg aumenta en 1.24 unidades, lo que sugiere en este modelo una relación positiva con la eficiencia.

- x2 = 2.8445. Un incremento en el desplazamiento aumenta el mpg en 2.84 unidades, reflejando también una asociación positiva poco usual.

- x3 = -2.3518. A medida que aumenta la potencia del motor, el mpg disminuye en 2.35 unidades, lo que indica un efecto negativo sobre la eficiencia.

- x4 = -0.5595. Valores más altos en la relación del eje trasero reducen el mpg en 0.56 unidades, mostrando una influencia negativa leve.

- x5 = -4.4764. Vehículos más pesados tienden a reducir el mpg en 4.48 unidades, confirmando que el peso extra empeora el rendimiento.

- x6 = -1.5293. Un incremento en el tiempo de aceleración en 1/4 de milla disminuye el mpg en 1.53 unidades, lo que sugiere que los autos más lentos resultan menos eficientes.

- x7 = 0.5587. Cuando el tipo de motor cambia, el mpg aumenta en 0.56 unidades, reflejando un efecto positivo aunque pequeño.

- x8 = 0.7584. Los vehículos con transmisión manual muestran un mejor rendimiento en 0.76 mpg, lo que implica una ligera mejora en la eficiencia.

- x9 = 2.2208. Un mayor número de marchas se relaciona con un incremento en el mpg de 2.22 unidades, indicando una influencia positiva en este modelo.

- x10 = -1.0744. Un mayor número de carburadores se asocia con una disminución del mpg en 1.07 unidades, reflejando una relación negativa con la eficiencia.

### Regularización L2 Test

In [38]:
l2 = Ridge(alpha = 0.1)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=0.1)

In [39]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.9008605095386456

In [40]:
l2.intercept_

np.float64(17.419097360139176)

In [41]:
l2.coef_

array([ 1.23746514,  2.28534641, -2.184281  ,  0.57670555, -4.03026084,
        1.33495812,  0.52588541,  0.71453978,  2.16983279, -1.32553793])

In [42]:
l2 = Ridge(alpha = 1)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=1)

In [43]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.8872304992805279

In [44]:
l2.intercept_

np.float64(17.383328961740464)

In [45]:
l2.coef_

array([ 0.62794712,  0.6879817 , -1.59000081,  0.47981821, -2.68884567,
        0.63051843,  0.36178546,  0.6206204 ,  1.53839604, -1.64658207])

In [46]:
l2 = Ridge(alpha = 10)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=10)

In [47]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.8278051382525741

In [48]:
l2.intercept_

np.float64(17.798139720722467)

In [49]:
l2.coef_

array([-0.49226589, -0.41981179, -0.76955512,  0.37221692, -1.31580654,
        0.11457072,  0.34950481,  0.6279019 ,  0.52828823, -0.9623776 ])

### Intepretación

Cuando se evalúa en el conjunto de prueba, se observa que con lambda = 0.1 el modelo logra el mejor ajuste, alcanzando un R2 de 0.901, lo que refleja una excelente capacidad predictiva, aunque con riesgo de sobreajuste por la baja penalización. Al aumentar a lambda = 1, el R2 baja ligeramente a 0.887, mostrando un balance más estable entre ajuste y generalización, gracias a la reducción de los coeficientes. Finalmente, con lambda = 10, el R2 desciende a 0.828, lo que indica que la penalización es demasiado fuerte, el modelo pierde poder explicativo, aunque gana simplicidad al tener coeficientes mucho más pequeños

## 1.2 Repite el ejercicio anterior usando 'qsec' como salida.

In [50]:
y = data["qsec"]
x = data.drop(columns=["qsec", "model"])

In [51]:
n = len(data)

In [52]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.4, random_state = 137)

## Escalamiento

In [53]:
from sklearn.preprocessing import StandardScaler

In [54]:
scaler = StandardScaler().fit(x_train)
X_train_scaled = scaler.transform(x_train)

In [55]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

### Train

In [56]:
y_hat = lr.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.978762192523052

In [57]:
n = len(X_train_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_train_scaled])
ols = sm.OLS(y_train, X)
results = ols.fit()
results.summary()

C:\Users\Nicolas\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=12 observations were given.
  return hypotest_fun_in(*args, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.349
Time:                        16:03:34   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.009      14.232      20.515
x1             1.0235      2.262      0.453      0.729     -27.713      29.760
x2            -0.3081     30.092     -0.010      0.993    -382.663     382.047
x3             0.2789      1.711      0.163      0.897     -21.465      22.023
x4            -1.0127      9.720     -0.104      0.934    -124.519     122.494
x5            -1.0356      8.615     -0.120      0.924    -110.497     108.426
x6             0.5976      1.295      0.461      0.725     -15.861      17.057
x7             0.2129     23.358      0.009      0.994    -296.578     297.004
x8            -0.3877     24.867     -0.016      0.990    -316.349     315.574
x9             0.6755     25.827      0.026      0.983    -327.485     328.836
x10           -0.1562      9.422     -0.017      0.989    -119.876     119.563
==============================================================================
Omnibus:                       12.359   Durbin-Watson:                   2.150
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.063
Skew:                           1.532   Prob(JB):                       0.0293
Kurtosis:                       5.178   Cond. No.                         517.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpretación Betas

- Const = 17.3733. Representa el tiempo base estimado en 1/4 de milla, cuando todas las variables predictoras son cero. El modelo estima alrededor de 17.37 segundos como punto de partida.

- x1 = 1.0235. A mayor número de cilindros, el tiempo en 1/4 de milla aumenta en 1.02 segundos, lo que indica que más cilindros se asocian con una aceleración más lenta.

- x2 = -0.3081. Un incremento en el desplazamiento del motor reduce el qsec en 0.31 segundos, sugiriendo que motores más grandes aceleran más rápido.

- x3 = 0.2789. A medida que aumenta la potencia del motor, el qsec sube en 0.28 segundos, lo que implica una aceleración ligeramente más lenta.

- x4 = -1.0127. Valores más altos en la relación del eje trasero disminuyen el qsec en 1.01 segundos, reflejando que una mayor relación de transmisión mejora la aceleración.

- x5 = -1.0356. Un aumento en el peso del vehículo reduce el qsec en 1.04 segundos, lo cual es contraintuitivo, ya que se esperaría lo contrario, pero en este modelo indica una aparente mejor aceleración en vehículos más pesados.

- x6 = 0.5976. Tiempos mayores se asocian con esta variable: un incremento en el predictor aumenta el qsec en 0.60 segundos, reflejando una aceleración más lenta.

- x7 = 0.2129. Cuando cambia el tipo de motor, el qsec se incrementa en 0.21 segundos, mostrando un efecto leve hacia una aceleración más lenta.

- x8 = -0.3877. Los vehículos con transmisión manual tienen un qsec menor en 0.39 segundos, lo que significa que aceleran más rápido que los automáticos.

- x9 = 0.6755. Un mayor número de marchas incrementa el qsec en 0.68 segundos, lo que sugiere que más marchas se asocian con una aceleración más lenta.

- x10 = -0.1562. Un mayor número de carburadores se asocia con una disminución del qsec en 0.16 segundos, lo que refleja una ligera mejora en la aceleración.

### Regularización L2 Train

In [58]:
l2 = Ridge(alpha = 0.1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=0.1)

In [59]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9773047022861614

In [60]:
l2.intercept_

np.float64(17.37333333333333)

In [61]:
l2.coef_

array([ 0.61143315, -0.24245986, -0.04895746, -0.71134739, -0.71445243,
        0.69741018,  0.47684547,  0.03899559,  0.06900507, -0.2456043 ])

In [62]:
l2 = Ridge(alpha = 1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=1)

In [63]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9671802657602481

In [64]:
l2.intercept_

np.float64(17.37333333333333)

In [65]:
l2.coef_

array([ 0.32243014, -0.423128  , -0.09911909, -0.48174143, -0.47038265,
        0.50302442,  0.45486151, -0.00142143, -0.19343426, -0.27205226])

In [66]:
l2 = Ridge(alpha = 1)  
l2.fit(X_train_scaled, y_train)

Ridge(alpha=1)

In [67]:
y_hat = l2.predict(X_train_scaled)
r2_score(y_train, y_hat)

0.9671802657602481

In [68]:
l2.intercept_

np.float64(17.37333333333333)

In [69]:
l2.coef_

array([ 0.32243014, -0.423128  , -0.09911909, -0.48174143, -0.47038265,
        0.50302442,  0.45486151, -0.00142143, -0.19343426, -0.27205226])

### Interpretación

Se observa que con lambda = 0.1 el modelo tiene el mejor ajuste R2 = 0.977, aunque con riesgo de depender demasiado de los predictores. Al aumentar a lambda = 1, el R2 baja ligeramente a 0.967, pero los coeficientes se estabilizan, lo que representa un mejor equilibrio entre ajuste y generalización. Finalmente, con lambda = 10, el R2 disminuye a 0.947, reflejando una penalización más fuerte que simplifica el modelo, aunque a costa de reducir su capacidad predictiva.

### Test

In [70]:
X_test_scaled = scaler.transform(x_test)

In [71]:
lr = LinearRegression()
lr.fit(X_test_scaled, y_test)

LinearRegression()

In [72]:
y_hat_test = lr.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.8836142852536799

In [73]:
n = len(X_test_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_test_scaled])
ols = sm.OLS(y_test, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.884
Model:                            OLS   Adj. R-squared:                  0.754
Method:                 Least Squares   F-statistic:                     6.833
Date:                Thu, 18 Sep 2025   Prob (F-statistic):            0.00399
Time:                        16:03:34   Log-Likelihood:                -17.875
No. Observations:                  20   AIC:                             57.75
Df Residuals:                       9   BIC:                             68.70
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.0020      0.351     48.425      0.000      16.208      17.796
x1             0.5791      0.601      0.964      0.360      -0.780       1.939
x2            -0.4788      0.693     -0.691      0.507      -2.047       1.089
x3            -1.1915      0.987     -1.207      0.258      -3.425       1.042
x4            -0.4364      0.824     -0.530      0.609      -2.301       1.428
x5             0.0103      0.367      0.028      0.978      -0.821       0.841
x6             1.7775      0.773      2.301      0.047       0.030       3.525
x7             0.2498      0.373      0.670      0.519      -0.593       1.093
x8            -0.6497      0.361     -1.800      0.105      -1.466       0.167
x9            -0.3125      0.600     -0.521      0.615      -1.669       1.044
x10           -0.4260      0.754     -0.565      0.586      -2.132       1.280
==============================================================================
Omnibus:                       17.263   Durbin-Watson:                   1.647
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.409
Skew:                           1.368   Prob(JB):                     5.01e-06
Kurtosis:                       7.670   Cond. No.                         26.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpretación betas

- Const = 17.0020. Representa el tiempo base en el cuarto de milla, que es de 17 segundos cuando todas las variables independientes son cero.

- x1 = -2.8811. A mayor número de cilindros, el qsec disminuye en 2.88 segundos, lo que significa que más cilindros se asocian con una aceleración más rápida.

- x2 = -13.3161. Un incremento en el desplazamiento del motor reduce el qsec en 13.32 segundos, sugiriendo que motores más grandes aceleran mucho más rápido.

- x3 = 13.4847. Cuando aumenta la potencia del motor, el qsec sube en 13.48 segundos, lo que indica una aceleración más lenta. Este resultado es contraintuitivo, porque se esperaría lo contrario.

- x4 = -16.6379. Valores más altos en la relación del eje trasero reducen el qsec en 16.64 segundos, reflejando que una mayor relación de transmisión favorece una aceleración más rápida.

- x5 = -3.1808. Un incremento en el peso del vehículo disminuye el qsec en 3.18 segundos, lo cual es inesperado, pues normalmente más peso debería ralentizar la aceleración.

- x6 = -11.1036. Cuando el tipo de motor cambia hacia recto (vs = 1), el qsec baja en 11.10 segundos, lo que sugiere que este tipo de motor se asocia con mayor rapidez en este modelo.

- x7 = -4.5950. Los vehículos con transmisión manual tienen un qsec menor en 4.60 segundos, lo que significa que aceleran más rápido que los automáticos.

- x8 = -3.7618. Un mayor número de marchas reduce el qsec en 3.76 segundos, asociándose con una aceleración más rápida.

- x9 = -6.4287. Un mayor número de carburadores disminuye el qsec en 6.43 segundos, lo que implica una mejora en la aceleración.

- x10 = 8.3255. Un mayor rendimiento de combustible incrementa el qsec en 8.33 segundos, mostrando que los autos más eficientes en consumo tienden a tener una aceleración más lenta

### Regularización L2 Test

In [74]:
l2 = Ridge(alpha = 0.1)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=0.1)

In [75]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.8828188950389997

In [76]:
l2.intercept_

np.float64(16.981672106209484)

In [77]:
l2.coef_

array([ 5.06349920e-01, -5.41049211e-01, -9.80458717e-01, -5.14656981e-01,
       -9.61998517e-04,  1.59536601e+00,  2.81794810e-01, -6.51761828e-01,
       -3.16486704e-01, -3.39567628e-01])

In [78]:
l2 = Ridge(alpha = 1)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=1)

In [79]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.8669489775816194

In [80]:
l2.intercept_

np.float64(16.97655668243855)

In [81]:
l2.coef_

array([ 0.23887246, -0.62032655, -0.42494672, -0.58382257, -0.00322778,
        0.99668557,  0.41839064, -0.64854303, -0.25409497, -0.20595853])

In [82]:
l2 = Ridge(alpha = 10)
l2.fit(X_test_scaled, y_test)

Ridge(alpha=10)

In [83]:
y_hat_test = l2.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.7635615962886475

In [84]:
l2.intercept_

np.float64(17.190796396487883)

In [85]:
l2.coef_

array([ 0.04216952, -0.4090824 , -0.12650283, -0.28664282,  0.00367278,
        0.31291366,  0.49948734, -0.45135553, -0.16689614, -0.20931837])

### Interpretación

Al evaluar en el conjunto de prueba, el modelo con lambda = 0.1 ofrece el mejor ajuste con un R2 de 0.882, mostrando buena capacidad predictiva aunque con riesgo de depender demasiado de los predictores. Con lambda = 1, el R2 baja levemente a 0.867, pero los coeficientes se vuelven más estables, ofreciendo un balance entre precisión y generalización. Finalmente, con lambda = 10, el R2 cae hasta 0.764, lo que evidencia que la penalización es demasiado fuerte, el modelo pierde poder explicativo aunque gana simplicidad.

## 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

In [86]:
data = pd.read_excel("Motor Trend Car Road Tests.xlsx")

In [87]:
y = data["mpg"]
x = data.drop(columns=["mpg", "model"])

In [88]:
x = pd.get_dummies(x, columns=["cyl", "gear", "carb"])

In [89]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.4, random_state = 137)

### Escalamiento

In [90]:
from sklearn.preprocessing import StandardScaler

In [91]:
scaler = StandardScaler().fit(x_train)
X_train_scaled = scaler.transform(x_train)

In [92]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

### Train

In [93]:
y_hat = lr.predict(X_train_scaled)
r2_score(y_train, y_hat)

1.0

In [94]:
n = len(X_train_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_train_scaled])
ols = sm.OLS(y_train, X)
results = ols.fit()
results.summary()

C:\Users\Nicolas\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=12 observations were given.
  return hypotest_fun_in(*args, **kwds)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\base\model.py:1527: RuntimeWarning: invalid value encountered in multip

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 18 Sep 2025   Prob (F-statistic):                nan
Time:                        16:03:35   Log-Likelihood:                 373.85
No. Observations:                  12   AIC:                            -723.7
Df Residuals:                       0   BIC:                            -717.9
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333        inf          0        nan         nan         nan
x1            -2.8807        inf         -0        nan         nan         nan
x2             1.1616        inf          0        nan         nan         nan
x3             3.8237        inf          0        nan         nan         nan
x4             2.0417        inf          0        nan         nan         nan
x5            -1.5490        inf         -0        nan         nan         nan
x6            -0.3475        inf         -0        nan         nan         nan
x7             0.5954        inf          0        nan         nan         nan
x8             1.6043        inf          0        nan         nan         nan
x9             0.0494        inf          0        nan         nan         nan
x10           -0.9860        inf         -0        nan         nan         nan
x11            1.4522        inf          0        nan         nan         nan
x12            0.1433        inf          0        nan         nan         nan
x13           -2.0876        inf         -0        nan         nan         nan
x14            2.2705        inf          0        nan         nan         nan
x15            0.0735        inf          0        nan         nan         nan
x16           -0.1647        inf         -0        nan         nan         nan
x17           -1.3885        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19           -0.4010        inf         -0        nan         nan         nan
==============================================================================
Omnibus:                        1.131   Durbin-Watson:                   1.923
Prob(Omnibus):                  0.568   Jarque-Bera (JB):                0.904
Skew:                           0.562   Prob(JB):                        0.636
Kurtosis:                       2.260   Cond. No.                         36.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

Los errores estándar salen como infinitos porque el modelo se ajusta demasiado bien a los datos, tiene muchas variables y pocas observaciones, además algunas variables están muy relacionadas entre sí. Esto hace que el modelo logre un ajuste perfecto R2 = 1, pero al mismo tiempo impide calcular de forma correcta la variabilidad de los coeficientes, por eso aparecen como inf y los valores t y p como nan.

### Interpretación betas

- Const = 17.4333. Representa el valor base del mpg cuando todas las variables son cero. Es el punto de partida del modelo.

- x1 = -2.8807. Esta categoría baja el rendimiento en 2.88 mpg.

- x2 = 1.1616. Esta categoría aumenta el rendimiento en 1.16 mpg.

- x3 = 3.8237. Aquí el rendimiento sube en 3.82 mpg, es un efecto positivo fuerte.

- x4 = 2.0417. El mpg aumenta en 2.04 mpg.

- x5 = -1.5490. Esta categoría reduce el mpg en 1.55 mpg.

- x6 = -0.3437. El mpg baja un poco, en 0.34 mpg.

- x7 = 0.5954. El rendimiento mejora ligeramente, sube en 0.60 mpg.

- x8 = 1.6043. Esta categoría hace que el mpg suba en 1.60 mpg.

- x9 = 0.0494. El efecto es casi nulo, apenas sube 0.05 mpg.

- x10 = -0.9860. Aquí el mpg baja en 0.99 mpg.

- x11 = 1.4522. El mpg se incrementa en 1.45 mpg.

- x12 = 0.1433. Aumenta muy poco, apenas 0.14 mpg.

- x13 = -2.0876. Reduce el rendimiento en 2.09 mpg.

- x14 = 2.2705. Esta categoría mejora el mpg en 2.27 mpg.

- x15 = 0.0735. El efecto es mínimo, sube apenas 0.07 mpg.

- x16 = -0.1647. El efecto es muy pequeño, baja en 0.16 mpg.

- x17 = -1.3885. Reduce el rendimiento en 1.39 mpg.

- x18 = 0.0000. No tiene efecto sobre el mpg.

- x19 = -0.4010. El mpg baja ligeramente en 0.40 mpg.

### Test

In [95]:
X_test_scaled = scaler.transform(x_test)

In [96]:
lr = LinearRegression()
lr.fit(X_test_scaled, y_test)

LinearRegression()

In [97]:
y_hat_test = lr.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.9311100085587481

In [98]:
n = len(X_test_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_test_scaled])
ols = sm.OLS(y_test, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.673
Method:                 Least Squares   F-statistic:                     3.604
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.112
Time:                        16:03:35   Log-Likelihood:                -35.912
No. Observations:                  20   AIC:                             103.8
Df Residuals:                       4   BIC:                             119.8
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         16.0206      1.461     10.969      0.000      11.965      20.076
x1             6.3920      6.138      1.041      0.356     -10.649      23.433
x2            -5.3342      4.128     -1.292      0.266     -16.796       6.127
x3             0.7169      1.780      0.403      0.708      -4.226       5.660
x4            -5.9789      4.282     -1.396      0.235     -17.869       5.911
x5             1.3110      2.266      0.579      0.594      -4.979       7.601
x6             1.2448      1.650      0.755      0.493      -3.336       5.825
x7             0.2750      2.217      0.124      0.907      -5.880       6.430
x8            -0.1980      2.136     -0.093      0.931      -6.129       5.733
x9            -0.8887      1.475     -0.602      0.579      -4.985       3.207
x10            0.9324      2.384      0.391      0.716      -5.686       7.551
x11           -1.5102      1.542     -0.979      0.383      -5.791       2.771
x12            0.2565      1.321      0.194      0.855      -3.410       3.923
x13            1.6998      1.427      1.191      0.300      -2.263       5.662
x14           -0.6891      2.884     -0.239      0.823      -8.697       7.319
x15           -0.9480      1.859     -0.510      0.637      -6.111       4.215
x16            0.7436      2.106      0.353      0.742      -5.103       6.591
x17           -0.5781      1.926     -0.300      0.779      -5.924       4.768
x18            1.8658      2.575      0.725      0.509      -5.283       9.014
x19           -4.8304      0.440    -10.969      0.000      -6.053      -3.608
==============================================================================
Omnibus:                       12.794   Durbin-Watson:                   2.135
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               13.060
Skew:                           1.155   Prob(JB):                      0.00146
Kurtosis:                       6.215   Cond. No.                     1.19e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.49e-36. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

### Interpretación

const = 16.0206. Representa el valor promedio base del rendimiento (mpg) cuando todas las demás variables son cero. Es decir, un punto de partida de 16.02 mpg.

x1 = 6.3920. Esta categoría aumenta el rendimiento en 6.39 mpg, un efecto claramente positivo.

x2 = -5.3342. Reduce el rendimiento en 5.33 mpg, mostrando una influencia negativa.

x3 = 0.7169. Incrementa ligeramente el mpg en 0.72 mpg, un efecto pequeño.

x4 = -5.9789. Disminuye el mpg en 5.98 mpg, mostrando un impacto negativo considerable.

x5 = 1.3110. Aumenta el mpg en 1.31 mpg, con un efecto positivo moderado.

x6 = 1.2448. Mejora el rendimiento en 1.24 mpg, otra influencia positiva.

x7 = 0.2750. Suma apenas 0.28 mpg, un efecto muy pequeño.

x8 = -0.1980. Baja un poco el mpg en 0.20 mpg, un efecto leve y negativo.

x9 = -0.8887. Reduce el rendimiento en 0.89 mpg, con una contribución negativa.

x10 = -1.9324. Disminuye el mpg en 1.93 mpg, reflejando un efecto desfavorable.

x11 = -1.5102. Baja el rendimiento en 1.51 mpg, con impacto negativo moderado.

x12 = 0.2565. Aumenta ligeramente el mpg en 0.26 mpg, un efecto casi nulo.

x13 = 1.6998. Incrementa el mpg en 1.70 mpg, un aporte positivo.

x14 = -0.6891. Reduce el mpg en 0.69 mpg, efecto pequeño y negativo.

x15 = -0.9480. Baja el rendimiento en 0.95 mpg, con leve efecto negativo.

x16 = 0.7436. Aumenta el mpg en 0.74 mpg, mejora ligera.

x17 = -0.5781. Reduce el rendimiento en 0.58 mpg, un efecto menor.

x18 = 1.8658. Mejora el mpg en 1.87 mpg, con influencia positiva moderada.

x19 = -4.8304. Tiene un efecto fuerte y negativo, reduce el mpg en 4.83 mpg.

## 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

In [99]:
data = pd.read_excel("Motor Trend Car Road Tests.xlsx")

In [100]:
y = data["qsec"]
x = data.drop(columns=["qsec", "model"])

In [101]:
x = pd.get_dummies(x, columns=["cyl", "gear", "carb"])

In [102]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.4, random_state = 137)

### Escalamiento

In [103]:
from sklearn.preprocessing import StandardScaler

In [104]:
scaler = StandardScaler().fit(x_train)
X_train_scaled = scaler.transform(x_train)

In [105]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

### Train

In [106]:
y_hat = lr.predict(X_train_scaled)
r2_score(y_train, y_hat)

1.0

In [107]:
n = len(X_train_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_train_scaled])
ols = sm.OLS(y_train, X)
results = ols.fit()
results.summary()

C:\Users\Nicolas\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=12 observations were given.
  return hypotest_fun_in(*args, **kwds)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
C:\Users\Nicolas\anaconda3\Lib\site-packages\statsmodels\base\model.py:1527: RuntimeWarning: invalid value encountered in multip

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 18 Sep 2025   Prob (F-statistic):                nan
Time:                        16:03:36   Log-Likelihood:                 359.72
No. Observations:                  12   AIC:                            -695.4
Df Residuals:                       0   BIC:                            -689.6
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733        inf          0        nan         nan         nan
x1            -2.8275        inf         -0        nan         nan         nan
x2             2.1520        inf          0        nan         nan         nan
x3            -0.1677        inf         -0        nan         nan         nan
x4             3.7193        inf          0        nan         nan         nan
x5             0.0248        inf          0        nan         nan         nan
x6             0.0007        inf          0        nan         nan         nan
x7            -0.0380        inf         -0        nan         nan         nan
x8             0.7525        inf          0        nan         nan         nan
x9             0.7933        inf          0        nan         nan         nan
x10           -1.1698        inf         -0        nan         nan         nan
x11            1.1786        inf          0        nan         nan         nan
x12            0.4526        inf          0        nan         nan         nan
x13           -2.0851        inf         -0        nan         nan         nan
x14            1.5119        inf          0        nan         nan         nan
x15           -0.1165        inf         -0        nan         nan         nan
x16            0.5565        inf          0        nan         nan         nan
x17           -1.8342        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19            0.6457        inf          0        nan         nan         nan
==============================================================================
Omnibus:                        6.788   Durbin-Watson:                   1.160
Prob(Omnibus):                  0.034   Jarque-Bera (JB):                2.863
Skew:                           0.892   Prob(JB):                        0.239
Kurtosis:                       4.594   Cond. No.                         113.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

Los errores estándar salen como infinitos porque el modelo se ajusta demasiado bien a los datos, tiene muchas variables y pocas observaciones, además algunas variables están muy relacionadas entre sí. Esto hace que el modelo logre un ajuste perfecto R2 = 1, pero al mismo tiempo impide calcular de forma correcta la variabilidad de los coeficientes, por eso aparecen como inf y los valores t y p como nan.

### Interpretación

- Const = 17.3733. Es el valor base del qsec cuando todas las demás variables son cero.

- x1 = -2.8275. Más cilindros reducen el qsec en 2.83 seg, indicando aceleración más rápida.

- x2 = 2.1520. Un mayor desplazamiento incrementa el qsec en 2.15 seg, sugiriendo aceleración más lenta.

- x3 = -0.1677. Mayor potencia del motor reduce el qsec en 0.17 seg, mejorando un poco la aceleración.

- x4 = 3.7193. Una relación del eje trasero mayor eleva el qsec en 3.72 seg, asociada con menor aceleración.

- x5 = 0.0248. Más peso aumenta apenas 0.02 seg el qsec, efecto mínimo.

- x6 = 0.0007. Otro efecto muy pequeño, casi nulo.

- x7 = -0.0380. Un motor recto reduce el qsec en 0.04 seg frente a uno en V.

- x8 = 0.7525. Transmisión manual aumenta el qsec en 0.75 seg, indicando aceleración más lenta.

- x9 = 0.7933. Más marchas aumentan el qsec en 0.79 seg, aceleración más lenta.

- x10 = -1.1698. Más carburadores reducen el qsec en 1.17 seg, aceleración más rápida.

- x11 = 1.1786. Esta categoría suma 1.18 seg al qsec.

- x12 = 0.4526. Incrementa en 0.45 seg el qsec.

- x13 = -2.0851. Reduce en 2.09 seg el qsec, mejor aceleración.

- x14 = 1.5119. Suma 1.51 seg al qsec.

- x15 = -0.1165. Resta 0.12 seg al qsec.

- x16 = 0.5565. Aumenta en 0.56 seg el qsec.

- x17 = -1.8342. Reduce el qsec en 1.83 seg.

- x18 = 0.0000. Sin efecto aparente (puede ser una dummy redundante).

- x19 = 0.6457. Incrementa el qsec en 0.65 seg.

### Test

In [108]:
X_test_scaled = scaler.transform(x_test)

In [109]:
lr = LinearRegression()
lr.fit(X_test_scaled, y_test)

LinearRegression()

In [110]:
y_hat_test = lr.predict(X_test_scaled)
r2_score(y_test, y_hat_test)

0.9085782920729732

In [111]:
n = len(X_test_scaled)
unos = np.ones([n,1])
X = np.column_stack([unos,X_test_scaled])
ols = sm.OLS(y_test, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.909
Model:                            OLS   Adj. R-squared:                  0.566
Method:                 Least Squares   F-statistic:                     2.650
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.179
Time:                        16:03:36   Log-Likelihood:                -15.460
No. Observations:                  20   AIC:                             62.92
Df Residuals:                       4   BIC:                             78.85
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         15.5191      0.497     31.239      0.000      14.140      16.898
x1             0.5580      0.964      0.579      0.594      -2.119       3.235
x2            -0.3350      2.483     -0.135      0.899      -7.230       6.560
x3             0.4043      1.756      0.230      0.829      -4.471       5.280
x4             0.0060      0.653      0.009      0.993      -1.807       1.819
x5             1.2050      1.779      0.677      0.535      -3.735       6.145
x6             0.0236      0.634      0.037      0.972      -1.737       1.784
x7            -0.8705      0.670     -1.300      0.264      -2.730       0.989
x8             0.5502      0.718      0.766      0.486      -1.444       2.544
x9             0.4040      0.516      0.783      0.477      -1.029       1.837
x10           -0.6937      0.802     -0.865      0.436      -2.920       1.532
x11           -0.0117      0.618     -0.019      0.986      -1.726       1.703
x12            0.3101      0.451      0.687      0.530      -0.943       1.563
x13           -0.3449      0.572     -0.603      0.579      -1.933       1.243
x14            0.6635      1.038      0.639      0.557      -2.217       3.544
x15            0.4220      0.695      0.607      0.576      -1.507       2.351
x16            0.6431      0.743      0.866      0.435      -1.419       2.705
x17            0.1190      0.700      0.170      0.873      -1.824       2.062
x18            0.6302      0.940      0.671      0.539      -1.979       3.239
x19           -4.6792      0.150    -31.239      0.000      -5.095      -4.263
==============================================================================
Omnibus:                        7.417   Durbin-Watson:                   1.968
Prob(Omnibus):                  0.025   Jarque-Bera (JB):                8.491
Skew:                           0.179   Prob(JB):                       0.0143
Kurtosis:                       6.172   Cond. No.                     7.64e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.36e-34. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

### Interpretación

- Const = 15.52. Representa el tiempo base promedio en 1/4 de milla (qsec) cuando todas las demás variables son cero.

- x1 = 0.5580. Más cilindros aumentan el tiempo en 0.56 segundos, haciendo la aceleración un poco más lenta.

- x2 = -0.3350. Un motor con mayor desplazamiento reduce el tiempo en 0.34 segundos, acelerando más rápido.

- x3 = 0.4043. Más potencia del motor aumenta el tiempo en 0.40 segundos, indicando aceleración más lenta.

- x4 = 0.0060. Efecto prácticamente nulo, solo cambia 0.006 segundos.

- x5 = 1.2050. Autos más pesados tardan 1.21 segundos más, acelerando más lento.

- x6 = 0.0236. Incremento muy pequeño, apenas afecta el tiempo.

- x7 = -0.8705. Los motores en línea reducen el tiempo en 0.87 segundos frente a los motores en V, mejorando la aceleración.

- x8 = 0.5502. Los autos con transmisión manual tardan 0.55 segundos más, acelerando un poco más lento.

- x9 = 0.4040. Más marchas aumentan el tiempo en 0.40 segundos, señal de aceleración más lenta.

- x10 = -0.6937. Más carburadores reducen el tiempo en 0.69 segundos, mejorando la aceleración.

- x11 = -0.0117. Efecto prácticamente nulo, apenas baja 0.01 segundos.

- x12 = 0.3101. Incrementa en 0.31 segundos el tiempo, aceleración un poco más lenta.

- x13 = -0.3449. Reduce el tiempo en 0.34 segundos, favoreciendo una aceleración más rápida.

- x14 = 0.6635. Aumenta el tiempo en 0.66 segundos, acelerando más lento.

- x15 = 0.4220. Incrementa el tiempo en 0.42 segundos, también más lento.

- x16 = 0.6431. Sube el tiempo en 0.64 segundos, efecto moderado hacia más lentitud.

- x17 = 0.1190. Efecto mínimo, apenas aumenta 0.12 segundos.

- x18 = 0.6302. Incrementa el tiempo en 0.63 segundos, aceleración más lenta.

- x19 = -4.6792. Disminuye el tiempo en 4.68 segundos, es decir, mejora mucho la aceleración.

## 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

El modelo del ejercicio 1.1, sin dummies, logra un ajuste bastante alto en el conjunto de entrenamiento R2 = 0.9901 y también un buen desempeño en prueba R2 = 0.9014. Sin embargo, al introducir las variables dummies en el ejercicio 2.1, el modelo alcanza un ajuste perfecto en entrenamiento R2 = 1.0, lo que sugiere que se adapta completamente a los datos de entrenamiento. En el conjunto de prueba, el R2 también mejora, pasando de 0.9014 a 0.9311, lo que indica que las variables categóricas aportan información adicional y ayudan a explicar mejor la variabilidad de mpg. En conclusión, el modelo con dummies presenta mejor capacidad predictiva tanto en entrenamiento como en prueba

## 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

En el ejercicio 1.2, sin incluir variables dummies, el modelo muestra un buen ajuste en entrenamiento R2 = 0.9788 y un desempeño aceptable en prueba R2 = 0.8836. Cuando se añaden las variables dummies en el ejercicio 2.2, el modelo mejora, en entrenamiento logra un ajuste perfecto R2 = 1.0, lo que indica que logra capturar toda la variabilidad de los datos de entrenamiento, mientras que en prueba también aumenta el desempeño R2 = 0.9086 frente a 0.8836. Esto sugiere que la inclusión de variables categóricas como cyl, gear y carb permite explicar mejor el tiempo en 1/4 de milla (qsec) y genera un modelo con mayor capacidad predictiva en comparación al caso sin dummies.